include/systems/eigen_system.h

```diff
@@ -179,61 +179,167 @@ class EigenSystem : public System
   static std::string precond_matrix_name() { return "Eigen Preconditioner"; }
 
   /**
-   * \returns The system matrix used for standard eigenvalue problems
+   * \returns A const reference to the system matrix used for standard eigenvalue problems.
+   *
+   * This matrix should only be available (and therefore this should only be called)
+   * if !_use_shell_matrices.
    */
-  inline SparseMatrix<Number> & get_matrix_A() const
+  inline const SparseMatrix<Number> & get_matrix_A() const
     {
-      libmesh_assert(matrix_A);
+      libmesh_assert(!_use_shell_matrices);
```

Comment:
This new assert looks good, but we should keep in the matrix_A assert if at all possible; same reason as I pointed out earlier.